.circleci/config.yml

```diff
@@ -0,0 +1,235 @@
+version: 2
+
+references:
+
+  # docker container for python only jobs
+  python_only_config: &python_only_config
+    working_directory: ~/marsha
+    docker:
+      - image: circleci/python:3.6.5
+        environment: &django_environ
+          DJANGO_SETTINGS_MODULE: marsha.settings
+          SECRET_KEY: MarshaSecretKey
+
+  # docker container for complete jobs: python+postgres
+  full_config: &full_config
+    working_directory: ~/marsha
+    docker:
+      - image: circleci/python:3.6.5
+        environment:
+          <<: *django_environ
+          DATABASE_URL: postgresql://root@localhost/marsha?sslmode=disable
+      - image: circleci/postgres:10.3
+        environment:
+          POSTGRES_USER: root
+          POSTGRES_DB: marsha
+
+  # build steps to save/restore the directory used by pip to cache downloaded packages
+  save_pip_cache: &save_pip_cache
+    save_cache:
+      key: v1-pip-cache-{{ .Branch }}-{{ .Revision }}
+      paths:
+        - ~/.cache/pip
+  restore_pip_cache: &restore_pip_cache
+    restore_cache:
+      keys:
+        - v1-pip-cache-{{ .Branch }}-{{ .Revision }}
+        - v1-pip-cache-{{ .Branch }}
+        - v1-pip-cache
+
+  # build steps to save/restore the virtualenv (to pass it between workflow steps, but also jobs)
+  save_venv: &save_venv
+    save_cache:
+      key: v2-venv-{{ .Branch }}-{{ .Revision }}
+      paths:
+        - ~/venv
+  restore_venv: &restore_venv
+    restore_cache:
+      keys:
+        - v2-venv-{{ .Branch }}-{{ .Revision }}
+        - v2-venv-{{ .Branch }}
+        - v2-venv
+
+  # build steps to save/restore the marsha repository directory (to pass it between workflow steps)
+  save_repo: &save_repo
+    save_cache:
+      key: v1-repo-{{ .Branch }}-{{ .Revision }}
+      paths:
+        - .
+  restore_repo: &restore_repo
+    restore_cache:
+      key: v1-repo-{{ .Branch }}-{{ .Revision }}
```

Comment:
I really don't want to do a new checkout at the beginning of each jobs. I can get that it may be faster, but I see it like "steps". First checkout. Then use this base to install code. Then use this new base to do the linters, etc...

I think we talk in durations less than a second or two here so we can ignore this at least for now.

Do you agree?

Reply:
I got your concerns. Note that I got this way of doing things from the... circle-ci frontend repository itself !! https://github.com/circleci/frontend/blob/master/.circleci/config.yml

About minimizing side effects, you get it wrong. There is no side effects because the directory is cached just after being checkouted / pip installe. And freshly restored as is each time.

So the only thing I agree with you is to not share venv betweens differents push-jobs (but keeping it between steps of a same push because of what I said above)

Reply:
Fair enough.